server/store.go

```diff
@@ -0,0 +1,840 @@
+	GetEditTypeSuggestions(user model.User) (badgesmodel.BadgeTypeList, error)
+	GetTypeSuggestions(user model.User) (badgesmodel.BadgeTypeList, error)
+	GetEditBadgeSuggestions(user model.User) ([]badgesmodel.Badge, error)
+
+	// API
+	AddBadge(badge badgesmodel.Badge) (*badgesmodel.Badge, error)
+	GrantBadge(badgeID badgesmodel.BadgeID, userID string, grantedBy string, reason string) (bool, error)
+	AddType(t badgesmodel.BadgeTypeDefinition) (*badgesmodel.BadgeTypeDefinition, error)
+	GetType(tID badgesmodel.BadgeType) (*badgesmodel.BadgeTypeDefinition, error)
+	GetBadge(badgeID badgesmodel.BadgeID) (*badgesmodel.Badge, error)
+	UpdateType(t badgesmodel.BadgeTypeDefinition) error
+	UpdateBadge(b badgesmodel.Badge) error
+	DeleteType(tID badgesmodel.BadgeType) error
+	DeleteBadge(bID badgesmodel.BadgeID) error
+
+	AddSubscription(tID badgesmodel.BadgeType, cID string) error
+	RemoveSubscriptions(tID badgesmodel.BadgeType, cID string) error
+	GetTypeSubscriptions(tID badgesmodel.BadgeType) ([]string, error)
+	GetChannelSubscriptions(cID string) ([]badgesmodel.BadgeTypeDefinition, error)
+
+	// PAPI
+	EnsureBadges(badges []badgesmodel.Badge, pluginID, botID string) ([]badgesmodel.Badge, error)
+
+	// DEBUG
+	DebugGetTypes() badgesmodel.BadgeTypeList
+}
+
+type store struct {
+	api plugin.API
+}
+
+func NewStore(api plugin.API) Store {
+	return &store{
+		api: api,
+	}
+}
+
+func (s *store) EnsureBadges(badges []badgesmodel.Badge, pluginID, botID string) ([]badgesmodel.Badge, error) {
+	l, err := s.getAllTypes()
+	if err != nil {
+		return nil, err
+	}
+
+	var tDef *badgesmodel.BadgeTypeDefinition
+	for i, t := range l {
+		if t.CreatedBy == botID {
+			tDef = &l[i]
+			break
+		}
+	}
+
+	if tDef == nil {
+		tDef, err = s.addType(badgesmodel.BadgeTypeDefinition{
+			Name:      "Plugin badges: " + pluginID,
+			CreatedBy: botID,
+		}, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	bb, err := s.getAllBadges()
+	if err != nil {
+		return nil, err
+	}
+
+	out := []badgesmodel.Badge{}
+	for _, pb := range badges {
+		found := false
+		for _, b := range bb {
+			if b.CreatedBy == botID && b.Name == pb.Name {
+				found = true
+				out = append(out, b)
+				break
+			}
+		}
+		if !found {
+			pb.Type = tDef.ID
+			pb.CreatedBy = botID
+			newBadge, err := s.AddBadge(pb)
+			if err != nil {
+				return nil, err
+			}
+			out = append(out, *newBadge)
+		}
+	}
+
+	return out, nil
+}
+func (s *store) DebugGetTypes() badgesmodel.BadgeTypeList {
+	l, _ := s.getAllTypes()
+	return l
+}
+
+func (s *store) AddBadge(b badgesmodel.Badge) (*badgesmodel.Badge, error) {
+	if !b.IsValid() {
+		return nil, errInvalidBadge
+	}
+
+	u, appErr := s.api.GetUser(b.CreatedBy)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	badgeTypes, err := s.getAllTypes()
+	if err != nil {
+		return nil, err
+	}
+
+	t := badgeTypes.GetType(b.Type)
+	if t == nil {
+		return nil, errors.New("missing badge type")
+	}
+
+	if !canCreateBadge(*u, *t) {
```

Comment:
I think we were cheking this at API level. I think checking the permissions should be at one specific level, we can check in the store, or we can check in the API, but checking it in both places is confusing.

This is just a recomendation, but it is not an stopper to get this merged.

Reply:
This one in particular (can create) we are checking only in the store. We are checking others here and there like can edit. Sometimes I am too wary, checking things twice in the same flow, but the main idea is to fail fast. For example, we are adding a sanity check before parsing the whole request of the edit badge dialog, or a check in the command before showing the edit dialog.

Reply:
Yes, I like the idea of having circles of confidence where you trust everything inside certain circle, if you have passed data through the API it should be valid data, and then, in the inside, it is all trusted data. I'm not against moving that validation to the store, actually that makes a lot of sense because you don't want to accidentally store inconsistent data in the store with your own logic. Anyway, the duplication of validation is what makes me uncomfortable.

Again, this is a general suggestion, not something to fix in this PR.